jpa/org.eclipse.persistence.jpa/src/org/eclipse/persistence/internal/jpa/deployment/PersistenceUnitProcessor.java

```diff
@@ -183,19 +183,61 @@ public static URL computePURootURL(URL pxmlURL, String descriptorLocation) throw
             // 210280: any file url will be assumed to always reference a file (not a directory)
             result = new URL(pxmlURL, path.toString()); // NOI18N
         } else if("jar".equals(protocol)) { // NOI18N
+            // JPA spec limits the possible scenarios in Java EE environments (8.2).
+            // However it makes sense to be a little more flexible, supporting
+            // any root URLs that resolve descriptorLocation to pxmlURL.
+            // For example, jar:file:/foo.jar!/bar/META-INF/persistence.xml
+            // will have root in jar:file:/foo.jar!/bar.
```

Comment:
As the reference implementation, I don't think I want to support any persistence.xml location not explicitly declared as part of section 8.2 of the JPA spec... I can understand wanting to be more flexible, but there is also the argument that ```jar:file:/foo.jar!/bar/META-INF/persistence.xml``` is just wrong according to the java jar specification. ```META-INF``` must be at the root of your jar. Users should understand the correct way of doing things rather than us trying to just accept anything.

Reply:
Ah, I forgot EclipseLink is a reference implementation. Then, indeed, extra flexibility is evil. I'll clean the comments/tests up tomorrow. That said, the logic itself shouldn't change, IMO - we still can have a persistence root in a directory inside a JAR. Validating the location of this root would be nice, but that's not the point of this PR.